### PR TITLE
Blob refactor

### PR DESCRIPTION
**Pull request by kirkeaton**
1. Changed Long variables to Int
2. Removed Java Buffer refinement
3. Modified Java Blob interface to replace Buffer with OutputStream and add a length parameter
4. Re-added InputStreamBlob
5. Modified FileBlob and ByteArrayBlob to be ProxyBlob's that delegate everything to InputStreamBlob
6. Modified RestBlobService & HttpBlobService to work with new interface.

### CODE REVIEW

File: src/foam/blob/HashingOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package foam.blob;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class HashingOutputStream
```
Comment (kgrgreer) [inline src/foam/blob/HashingOutputStream.java]: Add some docs.

File: src/foam/nanos/blob/HttpBlobService.java [hunk 2]
```diff
@@ -55,22 +55,19 @@ public void execute(X x) {
   }
 
   protected void download(HttpServletRequest req, HttpServletResponse resp) {
+    OutputStream os = null;
+
     try {
       String path = req.getRequestURI();
-      String id = path.substring(path.lastIndexOf("/") + 1);
+      String id = path.replaceFirst("/service/blobService/", "");
```
Comment (kgrgreer) [inline src/foam/nanos/blob/HttpBlobService.java]: Might someone ever want to mount this somewhere else? If so, you can find out where you're mounted by implementing NSpecAware so that you're provided your NSpec so that you can know where you're mounted. If not, then this is fine.

File: src/foam/blob/HashingOutputStream.java (hunk 1 above)
Comment (kirkeaton) [reply src/foam/blob/HashingOutputStream.java]: Done. Also added missing license header to HashingOutputStream and InputStreamBlob

File: src/foam/nanos/blob/HttpBlobService.java (hunk 2 above)
Comment (kirkeaton) [reply src/foam/nanos/blob/HttpBlobService.java]: Good point. Made the fix.

File: src/foam/blob/Blob.js [hunk 3]
```diff
@@ -22,13 +22,11 @@ foam.CLASS({
 
   properties: [
     {
-      class: 'Long',
+      class: 'Int',
```
Comment (adamvy) [inline src/foam/blob/Blob.js:25]: Switching these all to Ints will limit us to Blobs under 4GB will it not?  Why the change?

File: src/foam/blob/RestBlobService.java [hunk 4]
```diff
@@ -59,23 +59,19 @@ public Blob put_(X x, Blob blob) {
...
+      int chunk = 0;
+      int size = blob.getSize();
+      int chunks = (int) Math.ceil((double) size / (double) BUFFER_SIZE);
+
       while ( chunk < chunks ) {
-        buffer = blob.read(buffer, chunkOffset(chunk));
-        byte[] buf = buffer.getData().array();
-        os.write(buf, 0, (int) buffer.getLength());
-        buffer.getData().clear();
+        blob.read(os, chunk * BUFFER_SIZE, size);
```
Comment (adamvy) [inline src/foam/blob/RestBlobService.java:71]: "size" is the total size of the blob, but we're trying to read it in chunks of size BUFFER_SIZE.

The third parameter should be ( size % BUFFER_SIZE == 0 ? BUFFER_SIZE : size % BUFFER_SIZE ) or similar.

If size % BUFFER_SIZE is 0, then all chunks are the same size, if non-zero then it should give you the size of the final chunk.

File: src/foam/blob/Blob.js (hunk 3 above)
Comment (kirkeaton) [reply src/foam/blob/Blob.js:25]: I was noticing in a lot of cases previously that I was casting down to an integer anyway because the read & write functions of InputStream / OutputStream respectively don't take long's as the offset and length parameters. But beyond that do we ever envision having to upload a file larger than 4GB? If that's the case then we can modify this.

File: src/foam/blob/RestBlobService.java (hunk 4 above)
Comment (kirkeaton) [reply src/foam/blob/RestBlobService.java:71]: I think this can actually just be simplified to ```blob.read(os, 0, size);```. The read function can perform it's own internal chunking.

Comment (adamvy) [reply src/foam/blob/RestBlobService.java:71]: I agree

File: src/foam/blob/Blob.js (hunk 3 above)
Comment (adamvy) [reply src/foam/blob/Blob.js:25]: I have use cases in other work and at home where I'll want to deal with larger files :)
